### ISSUE TITLE
DockerTag does not accept !Ref

### ISSUE THREAD

**Issue by tiberiuscoman** [aws/aws-sam-cli#4474]
### Description:

DockerTag does not accept !Ref

### Steps to reproduce:

Init a serverless template with lambda using an Image.
Set Metadata:

Parameters:
  GitShaHash:
    Type: String
    Default: TEST
 
Metadata:
  Dockerfile: Dockerfile
  DockerContext: ./<function_code>
  DockerTag: !Ref GitShaHash

### Observed result:

In cdk.out template: 
Metadata:
  DockerContext: <path_to_lambda_code>
  DockerTag: GitShaHash
  Dockerfile: Dockerfile
  SamResourceId: <resource_id>

### Expected result:

Metadata:
  DockerContext: <path_to_lambda_code>
  DockerTag: Ref: GitShaHash
  Dockerfile: Dockerfile
  SamResourceId: <resource_id>
I am using the same parameter to tag the lambda function itself and it works.
### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: Windows 11
2. `sam --version`: 1.66.0
3. AWS region: eu-west-2

**Comment by jfuss** [comment]
@tiberiuscoman What command are you running?

**Comment by tiberiuscoman** [comment]
@jfuss  
sam build
sam deploy --profile <my_profile> --stack-name <my_stack_name> --capabilities CAPABILITY_IAM --parameter-overrides GitShaHash=TEST --template-file .aws-sam/build/template.yaml ----image-repository <my_ecr_repo>

**Comment by jfuss** [comment]
@tiberiuscoman `DockerTag` is used during `sam build`. If your want a docker image to be created with that tag, you need to add `--parameter-overrides GitShaHash=TEST` to the `sam build` command.

**Comment by tiberiuscoman** [comment]
Thanks @jfuss, it worked when I moved the --parameter-overrides to sam build.

**Comment by sachahu1** [comment]
I don't think the issue is closed though. While this works it sounds more like a hack to fix a bug than anything. I believe passing parameters through the command line defeats the purpose of having the `samconfig.toml

Could this be re-opened so the team can see it and fix the issue @tiberiuscoman

**Comment by tiberiuscoman** [comment]
@sachahu1 reopening it.

**Comment by sachahu1** [comment]
Thanks !

**Comment by jfuss** [comment]
@sachahu1 This is not a hack, this is the solution.

In the template, you are `Ref: CFNParameter` that is completely separate from `samconfig.toml` and the `samconfig.toml` is for parameters to the CLI NOT template. 

You can however put this into the `samconfig.toml` in the `parameter_overrides` key.

Closing as there are no actions.

**Comment by sachahu1** [comment]
Sorry @jfuss , maybe I was unclear. I did put the DockerTag as a parameter_overrides in the samconfig.toml and it is not replaced during the `sam build` step.

Here's an example of what I mean:
The cfn.yml template:
```
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31
Description: AWS CloudFormation template
Parameters:
  DockerImageTag:
    Type: String
    
Resources:
  ALambda:
    ...
    Metadata:
      DockerTag: !Ref DockerImageTag
      DockerContext: .
      Dockerfile: Dockerfile
```
The associated samconfig.toml:

```
[test]
[test.deploy]
[test.deploy.parameters]
stack_name = "test"
confirm_changeset = false
capabilities = ["CAPABILITY_IAM", "CAPABILITY_NAMED_IAM", "CAPABILITY_AUTO_EXPAND"]
parameter_overrides = [
    "DockerImageTag=test-tag-123"
]
image_repositories = [
    "ALambda=..."
]
```

My understanding is that even during the build step, if you run something like `sam build --config-file samconfig.toml --config-env test` it should replace the `DockerImageTag` parameter with the value `test-tag-123`. However, what ends up happening, in `.aws-sam`, is the `build.toml` has this:
```
[function_build_definitions.d2e414e6-5c41-4681-b3d6-de7feac48504.metadata]
DockerTag = "DockerImageTag"
DockerContext = "/home/sacha/..."
Dockerfile = "Dockerfile"
```
and the cloudformation templates built by sam (`.aws-sam/build/cfn.yml`) have the same information:

```
   ...
   
        DockerImageTag:
          Ref: DockerImageTag
```

In short, I agree that the parameter-overrides works on the CLI, I believe if you pass a `samconfig.toml` to the `sam build` command AND the `samconfig.toml` contains the `parameter_overrides` section, it should in fact behave the same as if you passed it directly to the CLI.

I hope that's a little clearer, maybe I'm still missing something but it feels like the CLI behaviour isn't consistent with the behaviour you get if you put your CLI arguments in a samconfig.toml instead.

**Comment by sachahu1** [comment]
@jfuss To add to above:
I created a minimal cloudformation template to make sure whether it is a bug. Ran the following two versions (which should achieve the same result).
I ran this command: `sam build --template-file cfn.yml  --parameter-overrides DockerImageTag=test`
and this `sam build --config-file samconfig.toml --template-file cfn.yml --config-env test` with this samconfig.toml:
```
version = 0.1

[test]
[test.deploy]
[test.deploy.parameters]
stack_name = "test"
confirm_changeset = false
capabilities = ["CAPABILITY_IAM", "CAPABILITY_NAMED_IAM", "CAPABILITY_AUTO_EXPAND"]
parameter_overrides = [
    "Environment=TEST",
    "DockerImageTag=test"
]
image_repositories = [
    "TemplateLambda=...",
]
```

Then ran a diff between the two `.aws-sam` folders:
```
Common subdirectories: .aws-sam/build and .aws-sam-invalid/build
diff .aws-sam/build.toml .aws-sam-invalid/build.toml
4c4
< [function_build_definitions.91d76fae-db95-4e51-add2-55389d469ba4]
---
> [function_build_definitions.e5da1d59-20a4-40a0-97a8-c9455bb443d6]
8,9c8,9
< [function_build_definitions.91d76fae-db95-4e51-add2-55389d469ba4.metadata]
< DockerTag = "test"
---
> [function_build_definitions.e5da1d59-20a4-40a0-97a8-c9455bb443d6.metadata]
> DockerTag = "DockerImageTag"
```

**Comment by sachahu1** [comment]
In case this ever ends up helping someone else:
I figured out my mistake, the `[test.build.parameters]` section of the `samconfig.toml` also allows for `parameter_overrides`.
So to correct the minimal example above, this `samconfig.toml` works:
```
version = 0.1

[test]
[test.build]
[test.build.parameters]
parameter_overrides = [
    "DockerImageTag=test"
]

[test.deploy]
[test.deploy.parameters]
stack_name = "test"
confirm_changeset = false
capabilities = ["CAPABILITY_IAM", "CAPABILITY_NAMED_IAM", "CAPABILITY_AUTO_EXPAND"]
parameter_overrides = [
    "Environment=TEST",
]
image_repositories = [
    "TemplateLambda=...",
]
```

**Comment by jfuss** [comment]
@sachahu1 `samconfig.toml` is based on commands, so `deploy` configuration is not read during `build`. You can either use `test.globals.parameters` or define properties for `test.build.parameters`. 

Docs: https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-config.html

**Comment by jfuss** [comment]
Ahh.. My browser was cached and looks like you solved it. The only difference is you can leverage `test.globals.parameters` to allow values to apply to multiple commands.

**Comment by sachahu1** [comment]
Thanks ! I had a quick look at the code and ended up figuring that one out :sweat_smile: Thanks again for your help here.
